### PR TITLE
xplat: Adds FreeBSD support

### PR DESCRIPTION
**Pull request by jasonwilliams200OK**
Refs: https://github.com/dotnet/corefx/pull/2068

R=@stephentoub,@mmitche,@josteink,@janhenke,@ghuntley

### CODE REVIEW

File: src/xunit.netcore.extensions/PlatformID.cs [hunk 1]
```diff
@@ -9,10 +9,11 @@ namespace Xunit
...
     {
-        Windows = 1,
+        Any = ~0,
+        AnyUnix = FreeBSD | Linux | OSX,
+        FreeBSD = 8,
         Linux = 2,
         OSX = 4,
-        AnyUnix = Linux | OSX,
-        Any = Windows | Linux | OSX
+        Windows = 1
```
Comment (stephentoub) [inline src/xunit.netcore.extensions/PlatformID.cs]: I'd prefer to leave this not sorted alphabetically.  If the list eventually grows long, we can consider it, but with so few items, I prefer the ordering to be the way it was, just with FreeBSD inserted between OSX and AnyUnix.

File: src/xunit.netcore.extensions/Interop.PlatformDetection.cs [hunk 2]
```diff
@@ -11,14 +11,15 @@ internal static partial class Interop
...
-        OSX
+        OSX,
+        Windows
     }
 
-    internal static bool IsWindows
+    internal static bool IsFreeBSD
     {
-        get { return OperatingSystem.Windows == PlatformDetection.OperatingSystem; }
+        get { return OperatingSystem.FreeBSD == PlatformDetection.OperatingSystem; }
```
Comment (stephentoub) [inline src/xunit.netcore.extensions/Interop.PlatformDetection.cs]: When will this return true?  You haven't modified PlatformDetection.OperatingSystem to support FreeBSD.

**Comment by stephentoub** [comment]
Did you build and test this locally to build corefx?  The CI system for buildtools only builds it; it doesn't then try to use the newly built build tools to build corefx.  We shouldn't take a PR in unless we know we at least haven't broken support for building corefx for the other platforms.

File: src/xunit.netcore.extensions/PlatformID.cs [hunk 3]
```diff
@@ -9,10 +9,11 @@ namespace Xunit
...
+        AnyUnix = FreeBSD | Linux | OSX,
         Windows = 1,
         Linux = 2,
         OSX = 4,
-        AnyUnix = Linux | OSX,
-        Any = Windows | Linux | OSX
+        FreeBSD = 8
     }
-}
+}
```
Comment (stephentoub) [inline src/xunit.netcore.extensions/PlatformID.cs]: The newline at the end of the file got removed.

File: src/xunit.netcore.extensions/Interop.PlatformDetection.cs (hunk 2 above)
Comment (jasonwilliams200OK) [inline src/xunit.netcore.extensions/Interop.PlatformDetection.cs]: I have added the code in `PlatformDetection.OperatingSystem` and amended the commit.

**Comment by Priya91** [comment]
Agree with @stephentoub, we need to test the latest buildtools with corefx build. Steps to test locally are outlined [here] (https://github.com/dotnet/buildtools/wiki/Local-testing-of-BuildTools-in-another-repo)

**Comment by jasonwilliams200OK** [comment]
@stephentoub, @Priya91, thanks for the review. 

I tested building it on Windows box as per the wiki instructions. On executing `"C:\Program Files (x86)\Microsoft Visual Studio 12.0\VC\vcvarsall.bat" x86 && build.cmd /p:SkipTests=true` (taken from https://github.com/dotnet/dotnet-ci/blob/master/jobs/dotnet_buildtools/config.xml#L52), it chokes with:

> CSC : error CS0518: Predefined type 'System.Object' is not defined or imported [C:\Users\Jason\Source\Repos\buildtools\
src\xunit.netcore.extensions\xunit.netcore.extensions.csproj]
>
    13 Warning(s)
    1 Error(s)
Time Elapsed 00:00:01.81
Build Exit Code = 1

Whereas when I execute only `build.cmd` (with which it uses VS 2015 RC on my box with both VS2013 and VS2015 installed), it throws 156 Errors. Among others there is this recurring one: "Predefined type 'System.Object' is not defined".

I have ditto result with `dotnet:buildtools` master branch, which could mean this is unrelated to my changes?

Can you confirm if there no other hack required to make it build?

**Comment by mmitche** [comment]
I believe this is a result of some newly published packages that are not supported with the current version of dnx.  My attempt to upgrade to a newer version yesterday hit a road-block.  Can you confirm whether any project.lock.json files got edited when you built using the new tools in corefx?

**Comment by jasonwilliams200OK** [comment]
@mmitche, with current master, it does not edit any lock file when I run:

```cmd
git clone https://github.com/dotnet/buildtools
cd build tools
"C:\Program Files (x86)\Microsoft Visual Studio 12.0\VC\vcvarsall.bat" x86 && build.cmd
:: or
"C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\vcvarsall.bat" x86 && build.cmd
```

It throws:

> 
CSC : error CS0518: Predefined type 'System.Object' is not defined or imported [C:\Users\Jason\Source\Repos\buildtools\src\xunit.netcore.extensions\xunit.netcore.extensions.csproj]
..
..
    13 Warning(s)
    1 Error(s)
Time Elapsed 00:00:17.53
Build Exit Code = 1

**Comment by saper** [comment]
This is what I get trying to build it using MSBuild xplat https://github.com/Microsoft/msbuild/commit/dc043c7bfad824f8ac8698eb3b18b4d9ec041185
using mono master.

**Comment by jasonwilliams200OK** [comment]
Thanks @saper, @akoeplinger for https://github.com/jasonwilliams200OK/buildtools/pull/1 and confirming it works on FreeBSD and Linux. :+1:

**Comment by saper** [comment]
Only nuget fails with:

````
Attempting to build package from 'Microsoft.DotNet.BuildTools.nuspec'.
File not found: 'Microsoft.DotNet.Build.Tasks\ECMA.snk'.
````

most probably a slash problem, because the files are there.

**Comment by jasonwilliams200OK** [comment]
@mmitche, these changes are tested on FreeBSD and Linux.

**Comment by mmitche** [comment]
LGTM

**Comment by mmitche** [comment]
I'll publish a new buildtools this afternoon.

**Comment by jasonwilliams200OK** [comment]
Thanks! :+1:

**Pull request merged**